### PR TITLE
Fix `MaxRootDomainInfoSpanningTree::computeInfoSibling`

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Edit: The original change of this PR is deprecated by https://github.com/csarofeen/pytorch/pull/2134 https://github.com/csarofeen/pytorch/pull/2111. So the majority of the discussion on this page is no longer related. But there is still one relatively small change that fixes `MaxRootDomainInfoSpanningTree::computeInfoSibling` worth having.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/test/test_gpu3.cpp [hunk 1]
```diff
@@ -6588,6 +6588,28 @@ TEST_F(NVFuserTest, FusionVectorizeRepro1843_CUDA) {
   testValidate(fusion, cg_outputs, {t1, t0}, {ref}, __LINE__, __FILE__);
 }
 
+TEST_F(NVFuserTest, FusionPersistentReduction_CUDA) {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/test/test_gpu3.cpp]: This test is unrelated to the change in this PR, I just feel that it is a good example to demonstrate normalization + reduction support.

**Comment by zasdfgbnm** [comment]
~Oh, please hold on. Unrequesting reviews. Need to look into something.~
Edit: ready for review now.

**Comment by naoyam** [comment]
The mirror is an interesting concept, and I see what you're trying to solve. I wonder if we could do something similar with IterDomain rather than TensorView. For the example above, the only information we would need to have is that the `b0` domains of `T1` and `T2` are exactly mapped. I'm guessing it would be conceptually simpler than having constraints with TensorViews.

**Comment by zasdfgbnm** [comment]
> The mirror is an interesting concept, and I see what you're trying to solve. I wonder if we could do something similar with IterDomain rather than TensorView. For the example above, the only information we would need to have is that the `b0` domains of `T1` and `T2` are exactly mapped. I'm guessing it would be conceptually simpler than having constraints with TensorViews.

Hmm, I don't know how that would work. For the above example
```
T0[I1, I2] = input
T1[b0, I1, I2] = broadcast(T0)
T2[b0, I1, I2] = broadcast(T0)
T3[b0, I1, r2] = sum(T1, axis=2)
T4[b0, I1, b2] = broadcast(T3)
T5[b0, I1, I2] = add(T2, T4)
T6[b0, I1, r2] = sum(T5, axis=2)
```
The `b0` in `T1` and `T2` is already exactly mapped based on the `T1-T3-T4-T5-T2` path. But this does not help transform propagation, because on the propagation path `T1->T0->T2`, when we propagate `T0->T2`, the transform propagator does not know the existence of `T1`, so it does not know how to order dimensions in `T2` correctly.

**Comment by naoyam** [comment]
Oh, they indeed appear to be mapped, but you also mentioned:

> After modification, the b0 in T1 and the b0 in T2 is not mapped.

**Comment by zasdfgbnm** [comment]
> Oh, they indeed appear to be mapped, but you also mentioned:
> 
> > After modification, the b0 in T1 and the b0 in T2 is not mapped.

Ahh, sorry for the confusion. I should be saying
> After modification, there is no way for the transform propagator to know that the `b0` in `T1` and the `b0` in `T2` are mapped when doing the propagation `T1->T0->T2`. As a result, a new propagation path is needed.

**Comment by zasdfgbnm** [comment]
> > Oh, they indeed appear to be mapped, but you also mentioned:
> > > After modification, the b0 in T1 and the b0 in T2 is not mapped.
> 
> Ahh, sorry for the confusion. I should be saying
> 
> > After modification, there is no way for the transform propagator to know that the `b0` in `T1` and the `b0` in `T2` are mapped when doing the propagation `T1->T0->T2`. As a result, a new propagation path is needed.

Looks like my entire statement about the first part is wrong. I have modified it. Hope it makes more sense this time.

**Comment by zasdfgbnm** [comment]
Oh, bad. The following example is still failing:
```C++
TEST_F(NVFuserTest, FusionPersistentReduction2_CUDA) {
  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
  auto fusion = fusion_ptr.get();
  FusionGuard fg(fusion);

  auto tv0 = makeContigTensor(2, DataType::Int);
  fusion->addInput(tv0);
  auto tv1 = broadcast(tv0, {true, false, false});
  auto tv2 = broadcast(tv0, {true, false, false});
  auto tv3 = sum(tv1, {2});
  auto tv4 = broadcast(tv3, {false, false, true});
  auto tv5 = add(tv2, tv4);
  auto tv6 = sum(tv5, {2});
  fusion->addOutput(tv6);

  auto options = at::TensorOptions().dtype(kFloat).device(at::kCUDA, 0);
  auto t0 = at::randn({1024, 768}, options).to(ScalarType::Long);
  auto ref = (t0 + t0.sum({1}, true)).sum({1}).unsqueeze(0);

  FusionExecutorCache fec(std::move(fusion_ptr));
  auto cg_outputs = fec.runFusionWithInputs({t0});

  auto runtime = fec.getMostRecentKernelRuntime();
  TORCH_INTERNAL_ASSERT(!runtime->isSegmented());

  testValidate(fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
}
```
The difference between this example and https://github.com/csarofeen/pytorch/issues/2094 is that, in the above example, I manually do a recompute, while in https://github.com/csarofeen/pytorch/issues/2094, the recompute is done by `RecomputeTv`...

**Comment by naoyam** [comment]
Thanks, I got it. Basically, as you mentioned, what we would want to do is having a recomputed tensor transformed like siblings to the original tensor. 

I'm a little concerned with the interface, though. We should generally try to minimize references between Vals as it's quite easy to corrupt references while transforming fusions.

Wouldn't it make more sense to have this mirroring information outside of TensorViews? It seems it's a scheduling property, i.e., some two tensors should be transformed in the same way, and I don't see why we would need to have such information embedded inside tensors. Since recompute is a scheduling primitive, the scheduler should know which recomputed tensor should be transformed like its original, and then that information could be passed to the transform propagator. This approach seems to look cleaner to me. What do you think?

**Comment by zasdfgbnm** [comment]
> Thanks, I got it. Basically, as you mentioned, what we would want to do is having a recomputed tensor transformed like siblings to the original tensor.
> 
> I'm a little concerned with the interface, though. We should generally try to minimize references between Vals as it's quite easy to corrupt references while transforming fusions.
> 
> Wouldn't it make more sense to have this mirroring information outside of TensorViews? It seems it's a scheduling property, i.e., some two tensors should be transformed in the same way, and I don't see why we would need to have such information embedded inside tensors. Since recompute is a scheduling primitive, the scheduler should know which recomputed tensor should be transformed like its original, and then that information could be passed to the transform propagator. This approach seems to look cleaner to me. What do you think?

Yes, I totally agree if we want to go to the mirroring direction.

But also, I just find a new example here https://github.com/csarofeen/pytorch/pull/2108#issuecomment-1284905033, and this example makes me feel that the "mirroring" approach is very narrow, and it only solves a few cases. Now I feel that we need to have a more generic transform propagation approach, but I don't yet know how.

**Comment by zasdfgbnm** [comment]
Maybe we should do the following: in `MaxInfoSpanningTree`, we should build a `ComputeAtRootDomainMap` and use it to find all direct-forwarding opportunities (including mirroring and sibling). And use direct-forwarding whenever possible on graph traversal.

**Comment by naoyam** [comment]
Don't we want to have hit fix rather quickly? If so, let's focus on the recompute case only now.

More generally, though, I wonder if we could just use the IterDomainGraph to do propagation without graph traversal. From IterDomainGraph, it should be possible to extract all IterDomain expressions applied to the reference tensor. Could we just replay those expressions on a given tensor?

**Comment by naoyam** [comment]
> Prefer the path that preserves concrete root id in MaxRootDomainInfoSpanningTree

Is there some technical reason about this? Or it just the error message being easier to understand as you mentioned? What does the error message look like?

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 2]
```diff
@@ -110,7 +110,12 @@ std::shared_ptr<PointwiseParams> getPointwiseHeuristics(
         (int64_t)dataTypeSize(inp->getDataType().value(), index_type));
   }
 
-  auto ref_root = largest_out->getMaybeRFactorDomain();
+  // We always cacheBefore output at the beginning of the scheduling. And after
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp]: Is this the fix for trivial reductions? Could you show an example with trivial reductions?

File: torch/csrc/jit/codegen/cuda/test/test_gpu3.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/test/test_gpu3.cpp]: So, this doesn't exercise the mirroring transformation? If not, we should probably add some unit tests.

**Review by naoyam — COMMENTED** [review]
The mirroring implementation in the propagator makes sense. The only concern I have for now is the change of TensorView as I mentioned.

**Comment by zasdfgbnm** [comment]
> Don't we want to have hit fix rather quickly?

Not for https://github.com/csarofeen/pytorch/issues/2094 on the codegen side. For https://github.com/csarofeen/pytorch/issues/2094, since the broadcast is at the beginning of the fusion, we can just kick it out of the fusion on the integration side. @jjsjann123 is already on it, and I think he already finished.

> More generally, though, I wonder if we could just use the IterDomainGraph to do propagation without graph traversal. From IterDomainGraph, it should be possible to extract all IterDomain expressions applied to the reference tensor. Could we just replay those expressions on a given tensor?

Hmm, this sounds ambitious but makes sense to me.

> Is there some technical reason about this? Or it just the error message being easier to understand as you mentioned? What does the error message look like?

It is just because it has 1 less step in debugging. It is not needed with the mirroring.

For the example
```
T0[I1, I2] = input
T1[b0, I1, I2] = broadcast(T0)
T2[b0, I1, I2] = broadcast(T0)
T3[b0, I1, r2] = sum(T1, axis=2)
T4[b0, I1, b2] = broadcast(T3)
T5[b0, I1, I2] = add(T2, T4)
T6[b0, I1, r2] = sum(T5, axis=2)
```

without this fix, the error message is

```
RuntimeError: Rfactor replay recieved an axis outside the number of dims in the tensor, acceptable inclusive range is -4 to 3
```

This is because we are propagating along the path `T3->T4->T5->T6` and the last dimension is not replayed. My first debug step is to make the propagation prefer the path `T3->T1->T0->T2->T5->T6`, which is the same path as a common non-failing normalization + reduction.

Then after this first debugging step, I start to see an indexing error, which is because `T6` is still not replayed correctly. Then I can check why `T3->T1->T0->T2->T5->T6` is not generating the correct replay as the second debug step.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp]: Yes, this is the fix for trivial reductions.

For example, if you have:
```
T0[b0, I1, I2] = input
T1[r0, I1, I2] = sum(T0)
output = T1
```
Then when computing the heuristics, all positions you are using are in `T1[r0, I1, I2]`. But during scheduling, you will do a `T1.cacheBefore()` and get
```
T0[b0, I1, I2] = input
T2[r0, I1, I2] = sum(T0)
T1[I1, I2] = set(T2)
output = T1
```
so all the positions you are referring to are now in `T1[I1, I2]`, which is 1 less than in `T1[r0, I1, I2]`. And this is another example why I hate trivial reduction and want to replace it with squeeze :).

**Comment by naoyam** [comment]
> > Don't we want to have hit fix rather quickly?
> 
> Not for #2094 on the codegen side. For #2094, since the broadcast is at the beginning of the fusion, we can just kick it out of the fusion on the integration side. @jjsjann123 is already on it, and I think he already finished.

OK, in either case, if the recompute + mirror approach is good enough at least for now, we should probably get this merged in.

> 
> > More generally, though, I wonder if we could just use the IterDomainGraph to do propagation without graph traversal. From IterDomainGraph, it should be possible to extract all IterDomain expressions applied to the reference tensor. Could we just replay those expressions on a given tensor?
> 
> Hmm, this sounds ambitious but makes sense to me.
> 
> > Is there some technical reason about this? Or it just the error message being easier to understand as you mentioned? What does the error message look like?
> 
> It is just because it has 1 less step in debugging. It is not needed with the mirroring.
> 
> For the example
> 
> ```
> T0[I1, I2] = input
> T1[b0, I1, I2] = broadcast(T0)
> T2[b0, I1, I2] = broadcast(T0)
> T3[b0, I1, r2] = sum(T1, axis=2)
> T4[b0, I1, b2] = broadcast(T3)
> T5[b0, I1, I2] = add(T2, T4)
> T6[b0, I1, r2] = sum(T5, axis=2)
> ```
> 
> without this fix, the error message is
> 
> ```
> RuntimeError: Rfactor replay recieved an axis outside the number of dims in the tensor, acceptable inclusive range is -4 to 3
> ```
> 

Let's fix the typo  (`recieved`)

> This is because we are propagating along the path `T3->T4->T5->T6` and the last dimension is not replayed. My first debug step is to make the propagation prefer the path `T3->T1->T0->T2->T5->T6`, which is the same path as a common non-failing normalization + reduction.
> 
> Then after this first debugging step, I start to see an indexing error, which is because `T6` is still not replayed correctly. Then I can check why `T3->T1->T0->T2->T5->T6` is not generating the correct replay as the second debug step.

Thanks for the explanation.

**Comment by zasdfgbnm** [comment]
I think we can put this PR on hold for a short time. I can split out the trivial reduction fix, and the typo fix (`recieved`), merge them first. And I will give try on transform propagation from ca map. If it turns out to be too hard, then we can come back to this mirroring approach.

**Comment by zasdfgbnm** [comment]
@naoyam The major portion of this PR is deprecated, so I removed these changes. But this PR contains a bug fix in `MaxRootDomainInfoSpanningTree::computeInfoSibling`, which still has its value. Please take a look.

**Review by naoyam — APPROVED** [review]
LGTM
